nemo_aligner/algorithms/supervised.py

```diff
@@ -134,10 +136,16 @@ def train_single_step(self, batch):
         return loss_mean, trainer_metrics | metrics
 
     def fit(self):
-        if self.cfg.max_epochs is not None and self.cfg.max_epochs > 1:
-            # because we need to figure out a nice way to reset the shuffling on our dataset
-            # otherwise epoch > 1 will loop over the dataset in the same order
-            raise ValueError("epoch > 1 is not supported")
+        if (not isinstance(self.train_dataloader.batch_sampler, MegatronPretrainingRandomBatchSampler)) and (
+            self.cfg.max_epochs is not None and self.cfg.max_epochs > 1
+        ):
+            # if you use MegatronPretrainingBatchSampler as the batch_sampler passed to your train dataloader (in builders.py)
+            # then each epoch will repeat all your samples in the same order as the previous epoch, there is no shuffling
+            # to fix this, you should use MegatronPretrainingRandomBatchSampler instead, which alleviates this issue and allows
```

Comment:
Yes, that is a good catch, I will look into this. I think we can just change it to True so long as we force the batch size to always be 1 it shouldn't change the downstream logic in the PPO classes from what it is now.

Reply:
Hmm, I'd rather avoid forcing `micro_batch_size=1` if possible, since if we can fit in a larger batch size then it's even better.

I actually don't know what `load_gbs` does -- I haven't looked at this sampler code yet. Would you mind adding some comment in the code that explains it? Hopefully once I understand I can think of potential solutions...

Reply:
I have changed `load_gbs=True` for PPO using the new sampler and it seems to work correctly with the existing PPO custom collate function. Can you or @gshennvm  please test it out on your side and see if it has any issues?

Reply:
It seems to me that this will break the rollout generation because of this loop that is supposed to iterate on each microboatch, but with `load_gbs=True` the batches will be larger than intended:

https://github.com/NVIDIA/NeMo-Aligner/blob/357875837ade48c974bda0678d60bff1133a4c08/nemo_aligner/algorithms/ppo.py#L210

I need to fix my dev setup to test it properly but let me know if I'm missing something.

Reply:
Tested it => can confirm `load_gbs=True` breaks PPO for the reason I mentioned above.

I can see two ways to address this:
1. Convert PPO to use a global batch sampler instead of a microbatch one
2. Make the sampler compatible with multi-epoch training

My preference would be (2) because the rollout generation logic in PPO is more naturally implemented with microbatch-based sampling. Doing (1) might be easier but would make the PPO code less readable and intuitive.

So for instance we could:
* As a first step, implement in NeMo-Aligner a version of `MegatronPretrainingRandomSampler` that supports multi-epoch training
* As a second step, either upstream this class into NeMo or modify `MegatronPretrainingRandomSampler` to add an option so that it supports multi-epoch training, then remove our custom class from NeMo-Aligner

This also makes me realize that we're now going to shuffle the data twice: once in the dataset class, and once in the sampler. This seems useless and potentially confusing (for instance someone may try to remove the shuffling from the dataset expecting the data to remain sorted) => I would suggest to get rid of the shuffling logic in the datasets, unless someone sees a good reason to keep it.

Reply:
we should drop the shuffling on the dataset level and use this sampler instead.

have to use load_gbs=False for PPO exactly as you mentioned in 1. because we do padding per microbatch and loading a gbs(I've tried doing this before) increases the padding dramatically. In theory we can do some sort of inflight batching but I rather leave load_gbs=False for now

Reply:
If we merge in this PR, then I can remove all the Dataset-side shuffling as part of my data layer rework PR.

I'll have a look at adding multi-epoch support to `MegatronPretrainingRandomSampler`

Reply:
Okay, I have added a custom Aligner-specific version of `MegatronPretrainingRandomSampler` and tied it into PPO. Can you please test it with max_epochs > 1 and see if it works now?

Reply:
Thanks, the code looks good to me, and seems to run fine locally. I'll just schedule a multi-epoch training run (with resume) to be sure it's fully working.

Reply:
So there's actually a bug that makes PPO crash (my first local run hadn't actually finished the first epoch so I hadn't seen it there). I believe the same issue is probably affecting other trainers but it might be silent for them.

The problem is that the step iterator defined at https://github.com/NVIDIA/NeMo-Aligner/blob/e4df57589f5ce008d84bef476925145f46390324/nemo_aligner/algorithms/ppo.py#L377 is using `total=self.max_steps` but this should actually be lower since it's supposed to represent the number of steps in a single epoch.

Btw in the PPO config can you please move `max_epochs: 1 # anything above 1 not supported` up above `max_steps` and remove that comment.

Reply:
But your max_steps should be set to the exact number of steps in a single epoch via the call to `set_max_steps()` which happens before the loop. Here is the code for `set_max_steps`:

```
def set_max_steps(self):
        self.max_steps = (self._train_dataloader_len * (self.cfg.max_epochs - self.epoch)) + self.step

        if (max_steps := self.cfg.get("max_steps", -1)) >= 0:
            self.max_steps = min(self.max_steps, max_steps)
```

The epoch length is derived from the `_train_dataloader_len` so it should correspond to a single epoch if you're starting a fresh run (it gets more complex if you resume one).

Reply:
Let's take a simple example where we start from scratch (`self.epoch == 0` and `self.step == 0`) and `max_steps` is not set in the config (so that the `if` condition is skipped). This gives
```
self.max_steps = _train_dataloader_len * self.cfg.max_epochs
```
which is the number of steps for all epochs, not a single one.

Reply:
Yes, that is correct, and by design. To do otherwise requires resetting self.step back to zero on each epoch, which has other problems, and creates a zig-zag pattern in WandB which doesn't look good.

In DPO/RM/SFT this is not a problem, because the inner loop tqdm is on the dataloader, but I see now in PPO it works differently and tqdms on the inner loop iter which is not good.

Reply:
It seems like this should fix it:

`global_pbar = tqdm(dataloader_iter, initial=self.step, total=self.max_steps, leave=True, desc="PPO Global Step")`

Reply:
I have pushed this fix, can you please try testing again?

Reply:
I see -- this code is quite confusing, ideally we'd simplify it but let's not worry about it for now...

There are a couple of issues with your suggested fix:
1. It would consume `dataloader_iter`
2. Although we could fix the first issue, because this loader works on micro-batches it still wouldn't work (the loop over `global_pbar` needs to be over rollout steps)

I have some idea of how to fix this and the other issue => let me give it a shot at implementing it, and if it works I can share the diff to get your feedback.

Reply:
So here's what I had in mind (this first commit is for this specific issue, I'll discuss the other part in the other thread): 70edf5ac

The main thing here is to ensure that `loop_iter` uses the correct number of steps to perform in the current epoch, so that we don't go beyond what's available in the dataloader.

This commit also fixes `max_steps` which I believe was incorrect for multi-epoch training because we were multiplyiing `_train_dataloader_len` by the number of epochs that remained, but this variable may take arbitrary values depending on how many samples have already been consumed in the current epoch (as you rightly pointed out).

I still need to test this more carefully, but please let me know if that looks right to you.

Reply:
> we should drop the shuffling on the dataset level and use this sampler instead.

Just realized this still needs to be done. Ideally would be in this PR so we don't break the shuffling logic twice in a row. If not trivial can be done in a subsequent PR, but let's try to get it in fast then. @trias702 thoughts?

Reply:
> we should drop the shuffling on the dataset level and use this sampler instead.

ah just realized this PR will actually break all our CI tests because our CI tests enforce a strict ordering of the losses since we check against the values. So it'd be really good if we can do this too to not change the data ordering again, i need to rerun all our baselines

Reply:
> i need to rerun all our baselines

Based on my own experience maintaining CI with exact loss match, you'll probably want a script to do that for you automatically because you will need it again and the more tests you have, the more annoying it is ;)

Reply:
I can edit all of the dataset classes to remove shuffling there for this PR?

Reply:
> I can edit all of the dataset classes to remove shuffling there for this PR?

That'd be great, thanks! Then I believe we can merge once CI is adjusted accordingly.

Reply:
> Then I believe we can merge once CI is adjusted accordingly.

Yup! turns out I just need to edit some stuff to get the CI to auto-regenerate the baselines :)

Reply:
Okay, I have edited all of the Aligner-side dataset classes to no longer do shuffling. However, I have kept the seed arg being passed to the constructor as I already have a big data rework PR which cleans up all those superfluous args and I will add seed there to keep it consistent.

Also note that I cannot remove shuffling on the GPTSFT dataset classes as that code is inside nemo, and there's no bool to turn it off which I can pass from the Aligner side. However, looking at the source code for those GPTSFT dataset classes, they only do shuffling when you pass `max_num_samples is not None` which looking at our code for `train_gpt_sft.py` we only pass that when max_steps >= 0. So if you use max_steps, then you will have dataset shuffling for the GPTSFT classes which is not possible to deactivate as there's no bool flag.

Reply:
Thanks @trias702! I believe even more code could be removed, in the interest of time I did a pass, can you please have a look at a7ffa57b and cherry-pick it if it looks good to you?
(I'm running a quick debug test with RM training & PPO to make sure there's no obvious issue at least for those)

Reply:
That's all done now, thanks